### ISSUE TITLE
running Git LFS status in parallel in the same repo reports index errors

### ISSUE THREAD

**Issue by robert914** [git-lfs/git-lfs#4665]
**Describe the bug**

If you use the `pre-commit` hook `check-added-large-files` found here: https://github.com/pre-commit/pre-commit-hooks
This hook will cause calling of command `git lfs status --json` multiple times in parallels for different file sets in the same repo due to how `pre-commit` processes many files in parallel to improve speed. If you run `git lfs status --json` multiple times at the same time in the same repo, it will result in errors. This causes that hook to incorrectly get the list of files managed by Git LFS because of this error. It was reported to the author here: https://github.com/pre-commit/pre-commit-hooks/issues/656

The author claims this is a bug with Git LFS and said to report it here.

**To Reproduce**

Sample script:
```bash
#!/bin/bash
git lfs status --json >run1.log &
git lfs status --json >run2.log &
git lfs status --json >run3.log &
git lfs status --json >run4.log &
```

Output received:
```
Failed to run git update-index: error running /usr/libexec/git-core/git 'update-index' '-q' '--refresh': '' 'exit status 128'
Failed to run git update-index: error running /usr/libexec/git-core/git 'update-index' '-q' '--refresh': '' 'exit status 128'
Failed to run git update-index: error running /usr/libexec/git-core/git 'update-index' '-q' '--refresh': '' 'exit status 128'
Failed to run git update-index: error running /usr/libexec/git-core/git 'update-index' '-q' '--refresh': '' 'exit status 128'
```

**Expected behavior**

No sure, should it error out? or should it actually work and return the files? If this is a invalid use pattern, then the author of the hook will need to fix the hook to not run it this way, but I cannot find documentation stating that parallel execution of `git lfs status` in a repo is not allowed.

**System environment**

Tested on CentOS 7 and MacOS.

**Output of `git lfs env`**

```
❯ git lfs env
git-lfs/3.0.1 (GitHub; linux amd64; go 1.17)
git version 2.24.4

Endpoint=https://github.com/xxx/xxx.git/info/lfs (auth=none)
  SSH=git@github.com:xxx/xxx.git
LocalWorkingDir=/drv/data/xxx/xxx
LocalGitDir=/drv/data/xxx/xxx/.git
LocalGitStorageDir=/drv/data/xxx/xxx/.git
LocalMediaDir=/drv/data/xxx/xxx/.git/lfs/objects
LocalReferenceDirs=
TempDir=/drv/data/xxx/xxx/.git/lfs/tmp
ConcurrentTransfers=8
TusTransfers=false
BasicTransfersOnly=false
SkipDownloadErrors=false
FetchRecentAlways=false
FetchRecentRefsDays=7
FetchRecentCommitsDays=0
FetchRecentRefsIncludeRemotes=true
PruneOffsetDays=3
PruneVerifyRemoteAlways=false
PruneRemoteName=origin
LfsStorageDir=/drv/data/xxx/xxx/.git/lfs
AccessDownload=none
AccessUpload=none
DownloadTransfers=basic,lfs-standalone-file,ssh
UploadTransfers=basic,lfs-standalone-file,ssh
GIT_EXEC_PATH=/usr/libexec/git-core
git config filter.lfs.process = "git-lfs filter-process"
git config filter.lfs.smudge = "git-lfs smudge -- %f"
git config filter.lfs.clean = "git-lfs clean -- %f"
```

**Comment by bk2204** [comment]
Hey,

I'm not able to reproduce on my system, which is a Debian unstable system, either with Git 2.20 or with a branch from `next`.  However, looking at this, I don't think there's a way we can avoid this.  We use `git diff-index` under the hood, and the only way that's going to work normally is for us to have done `git update-index --refresh`.  Even with `git update-index`, we only fail if we needed to write the index because something changed and we can't, so in this case, we really do need to fail here since the index legitimately needed to be updated but couldn't be.  `git status` will only write to the index if it's not otherwise contended, but `git update-index` won't do that.

Probably the easiest way to solve this problem is to avoid having the pre-commit hook modify the working tree.  Otherwise, unless Git learns some technique for performing an optional update (which it's possible I may write at some point), there's no way around this, and pre-commit will need to run the command once and cache it.  I'll mark this as an enhancement as a place we could potentially improve.

**Comment by danmcmahill** [comment]
As a note, in a small repo, this may or may not show up.  I think you need a big repo and possibly one with staged files that will end up in LFS.  For example I could not reproduce on the `git-lfs` repo but that is moderately small but could easily reproduce on a fairly large repo that contains quite a number of files in LFS.

**Comment by robert914** [comment]
Hi @bk2204, just for verification, @danmcmahill is correct, you may not see this unless you have a large repo with many files (both non-LFS and LFS). It tends to happen when the `git lfs status` (which calls `git update-index`) takes long enough for them to step on each other.

Thanks for adding this to the enhancement list. It would also be good to know if you recommend a replacement that should instead be used to fix the hook. I was wondering of just a `git lfs ls-files --name-only` would be a better replacement that wouldn't have this side effect of hitting the index to update it and fail? I wasn't sure if that would output the proper result if the file was newly added (i.e. not yet committed) vs. one that was already committed.

We are currently forcing the hook to run serially for the LFS check currently to get around this for now.  Unless a recommendation comes from you indicating that what is currently being done in that hook is not the proper way to do it, I don't know that the hook's author is going to be very willing to change the hook.

If necessary, we can always fork the hook code and make the recommended change ourselves to correct this problem properly.

Thanks!

**Comment by bk2204** [comment]
If you want to know which files are LFS files, the best way is to ask Git that question yourself by doing this:

```
$ git ls-files -z | git check-attr --stdin -z filter
```

If you get `filter` set to `lfs`, then it's an LFS file.  That doesn't involve asking Git LFS at all, and it will mean that the hook will work even if the user doesn't have Git LFS installed, which is probably better for most projects, since a common misconfiguration is to not install Git LFS.  You can also use `git lfs ls-files --name-only`, which I believe doesn't refresh the index.

**Comment by robert914** [comment]
@bk2204 Thanks for the information. I tested the fix by replacing the `lfs_files()` function in that hook using the `git lfs ls-files --name-only` method as a test (i.e. easier to implement for now). I've verified that the problem has now been eliminated by using that instead of the `git lfs status --json` method.

I was going to submit a fix suggestion to the hook author, but apparently I've been banned from creating issues in that repo due to the issues I submitted on this problem to him in the first place. So I guess we'll make our own version of the hook using your suggestions.

Thanks for the help!
